Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 6fa2f764eeb6edd510a346d8162c9ef5012a248c
Number of Error(s): 4
/mnt/disk1/home/vamosax/smoketest/PR-11036/HPCC-Platform/dali/base/dasess.cpp:2072:66: error: \xe2\x80\x98staticDigitalSignatureManagerInstance\xe2\x80\x99 was not declared in this scope

[Build error:](https://gist.github.com/52a2b82e33ad5d62d435a2e4e8ac367c)
Build log [Build log](https://gist.github.com/HPCCSmoketest/52a2b82e33ad5d62d435a2e4e8ac367c#file-result-18-04-17-13-11-40-log)

--- a/dali/base/CMakeLists.txt
+++ b/dali/base/CMakeLists.txt
@@ -64,6 +64,7 @@ include_directories (
          ./../../system/jlib 
          ./../../rtl/include 
          ./../../system/security/shared
+         ./../../system/security/digisign
     )
 
 ADD_DEFINITIONS( -DLOGMSGCOMPONENT=3 -D_USRDLL -DDALI_EXPORTS )
@@ -78,6 +79,7 @@ if(NOT PLUGIN)
     target_link_libraries(
         dalibase 
         jlib
+        digisign
         mp 
         hrpc 
         remote)

--- a/dali/base/dasess.cpp
+++ b/dali/base/dasess.cpp
@@ -33,6 +33,7 @@
 #include "daldap.hpp"
 #include "seclib.hpp"
 #include "dasess.hpp"
+#include "digisign.hpp"
 
 #ifdef _MSC_VER
 #pragma warning (disable : 4355)
@@ -602,7 +603,7 @@ public:
                 unsigned auditflags = 0;
                 if (mb.length()-mb.getPos()>=sizeof(auditflags))
                     mb.read(auditflags);
-                udesc->deserializeExtra(mb);//deserialize sessionToken and user signature if present (hpcc ver 7.0.0 and newer)
+                udesc->deserializeExtra(mb);//deserialize user signature if present
                 int err = 0;
                 SecAccessFlags perms = manager.getPermissionsLDAP(key,obj,udesc,auditflags,&err);
                 mb.clear().append((int)perms);
@@ -911,7 +912,7 @@ public:
 #endif
         udesc->serialize(mb);
         mb.append(auditflags);
-        udesc->serializeExtra(mb);//serialize sessionToken and user signature if Dali version
+        udesc->serializeExtra(mb);//serialize user signature if Dali version >= 3.14
         if (!queryCoven().sendRecv(mb,RANK_RANDOM,MPTAG_DALI_SESSION_REQUEST,SESSIONREPLYTIMEOUT))
             return SecAccess_None;
         SecAccessFlags perms = SecAccess_Unavailable;
@@ -2010,13 +2011,13 @@ class CUserDescriptor: implements IUserDescriptor, public CInterface
 {
     StringAttr username;
     StringAttr passwordenc;
-    unsigned sessionToken;//ESP session token
+    unsigned sessionToken = 0;//ESP session token
     StringBuffer signature;//user's digital Signature
+    IDigitalSignatureManager * pDSM = nullptr;
 public:
     IMPLEMENT_IINTERFACE;
     CUserDescriptor()
     {
-        sessionToken = 0;
     }
     StringBuffer &getUserName(StringBuffer &buf)
     {
@@ -2062,7 +2063,16 @@ public:
     void serializeExtra(MemoryBuffer &mb)
     {
         if (queryDaliServerVersion().compare("3.14") >= 0)
-            mb.append(sessionToken).append(signature.length()).append(signature);
+        {
+            if (signature.isEmpty())
+            {
+                if (pDSM == nullptr)
+                    pDSM = createDigitalSignatureManagerInstanceFromEnv();
+                if (pDSM && pDSM->isDigiSignerConfigured())
+                    pDSM->digiSign(username, signature);//Set user's digital signature
+            }
+            mb.append(signature);
+        }
     }
     void deserialize(MemoryBuffer &mb)
     {
@@ -2071,22 +2081,10 @@ public:
     void deserializeExtra(MemoryBuffer &mb)
     {
         signature.clear();
-        sessionToken = 0;
-        if (mb.remaining() > 0)
+        if (queryDaliServerVersion().compare("3.14") >= 0)
         {
-            mb.read((unsigned &)sessionToken);
-
             if (mb.remaining() > 0)
-            {
-                unsigned len = 0;
-                mb.read(len);
-                if (len)
-                {
-                    StringBuffer sig;
-                    mb.read(sig);
-                    signature.append(sig);
-                }
-            }
+                mb.read(signature);
         }
     }
 };

--- a/dali/server/CMakeLists.txt
+++ b/dali/server/CMakeLists.txt
@@ -38,6 +38,7 @@ include_directories (
          ./../../system/include 
          ./../../system/security/shared 
          ./../../system/security/LdapSecurity 
+         ./../../system/security/digisign
          ./../../system/jlib 
          ${CMAKE_BINARY_DIR}
          ${CMAKE_BINARY_DIR}/oss
@@ -60,6 +61,7 @@ target_link_libraries ( daserver
          hrpc 
          remote 
          dalibase 
+         digisign
     )
 IF (USE_OPENLDAP)
 target_link_libraries ( daserver LdapSecurity )

--- a/dali/server/daldap.cpp
+++ b/dali/server/daldap.cpp
@@ -25,6 +25,7 @@
 #include "daldap.hpp"
 #include "mpbase.hpp"
 #include "dautils.hpp"
+#include "digisign.hpp"
 
 #ifndef _NO_LDAP
 #include "seclib.hpp"
@@ -49,6 +50,7 @@ class CDaliLdapConnection: implements IDaliLdapConnection, public CInterface
     StringAttr              filesdefaultuser;
     StringAttr              filesdefaultpassword;
     unsigned                ldapflags;
+    IDigitalSignatureManager * pDSM = nullptr;
 
     void createDefaultScopes()
     {
@@ -141,9 +143,30 @@ public:
 
         Owned<ISecUser> user = ldapsecurity->createUser(username);
         user->credentials().setPassword(password);
-        if (!ldapsecurity->authenticateUser(*user, NULL))
+
+        //Check user's digital signature, if present
+        bool authenticated = false;
+        if (!isEmptyString(udesc->querySignature()))
         {
-            ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s fails authentication",key?key:"NULL",obj?obj:"NULL",username.str());
+            if (nullptr == pDSM)
+                pDSM = createDigitalSignatureManagerInstanceFromEnv();
+            if (pDSM && pDSM->isDigiVerifierConfigured())
+            {
+                StringBuffer b64Signature(udesc->querySignature());
+                if (!pDSM->digiVerify(username, b64Signature))//digital signature valid?
+                {
+                    ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s invalid user digital signature",key?key:"NULL",obj?obj:"NULL",username.str());
+                    return SecAccess_None;//deny
+                }
+                else
+                    authenticated = true;
+            }
+            user->credentials().setSignature(udesc->querySignature());
+        }
+
+        if (!authenticated && !ldapsecurity->authenticateUser(*user, NULL))
+        {
+            ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s fails LDAP authentication",key?key:"NULL",obj?obj:"NULL",username.str());
             return SecAccess_None;//deny
         }
 
@@ -205,12 +228,37 @@ public:
         udesc->getUserName(username);
         udesc->getPassword(password);
         Owned<ISecUser> user = ldapsecurity->createUser(username);
-        user->credentials().setPassword(password);
-        if (!ldapsecurity->authenticateUser(*user, &superUser) || !superUser)
+
+        //Check user's digital signature, if present
+        bool authenticated = false;
+        if (!isEmptyString(udesc->querySignature()))
         {
-            *err = -1;
-            return false;
+            if (nullptr == pDSM)
+                pDSM = createDigitalSignatureManagerInstanceFromEnv();
+            if (pDSM && pDSM->isDigiVerifierConfigured())
+            {
+                StringBuffer b64Signature(udesc->querySignature());
+                if (!pDSM->digiVerify(username, b64Signature))//digital signature valid?
+                {
+                    ERRLOG("LDAP: enableScopeScans(%s) : Invalid user digital signature", username.str());
+                    *err = -1;
+                    return false;
+                }
+                else
+                    authenticated = true;
+            }
         }
+
+        if (!authenticated)
+        {
+            user->credentials().setPassword(password);
+            if (!ldapsecurity->authenticateUser(*user, &superUser) || !superUser)
+            {
+                *err = -1;
+                return false;
+            }
+        }
+
         unsigned flags = getLDAPflags();
         if (enable)
         {

